### DISCUSSION TITLE
Tmux-style attach/reattach - Implementation questions

### DISCUSSION THREAD

**Discussion by LunarLambda** [neovim/neovim#30314] [General]
I'm trying to implement https://github.com/neovim/neovim/issues/23093 as a Lua plugin.

The implementation so far can be found here: https://git.buny.computer/lambda/nmux/src/branch/main/nmux.lua

The steps are as follows:

1. We compute the per-user, per-session server socket path. This is something like `/run/user/1000/nmux/default`.
2. We check whether a Neovim server is already listening on that socket. If yes, the script `exec`s itself into a `nvim --remote-ui --server <socket>` client that attaches to the server.
3. If not, we unlink the socket (if it existed, workaround https://github.com/neovim/neovim/issues/30123)
4. We spawn a new `nvim --headless --listen <socket>` process as the server, then re-`exec` the client as above.

This all works so far. We can spawn a server on-demand and connect to it from a new client or re-attach a client to an existing server.

Screen recording with the rough proof of concept: https://asciinema.org/a/8ptwM9Vm0xbI3jjy9q6o1ne9t

Now the difficult part. How can a client detach itself? I cannot figure out any way for a client to detach only itself in a way that is unambiguous and error-free. Previous attempts by other people simply detached all TUIs.

I'm not aware of any way that Lua code, which all runs inside of the server process and never the client process, can unambiguously figure out it's channel ID so it can call `chanclose()`, which is currently the only way to detach a remote-ui client without closing the server. Calling `nvim_detach_ui` is only possible via RPC, which, again, requires obtaining a channel ID.

**Comment by justinmk** [comment]
> How can a client detach itself?

Specifically you mean the *nvim builtin UI client*. Any other client can call `nvim_get_chan_info(0)` to get its channel id. But the `nvim` builtin UI client doesn't expose a way for users/plugins to execute code from its "local" context.

For the default case where the nvim ui spawns the server as a *child* process, I think it should work if (1) the `nvim` server is spawned with `detach=true`, (2) terminating the *parent* (ui) process should allow the *child* (server) to continue running.

For the non-default case (`nvim --remote-ui`), there are some obvious potential workarounds, such as storing a flag (`nvim --server ./hello.sock --remote-send ':let ui=42<cr>' && nvim --remote-ui ...`).

I see least two clear next steps:

1. `nvim +cmd --remote-ui` should execute `+cmd` in the remote server. No reason for it to behave the way it currently does.
2. For the builtin nvim TUI, "current UI" has no mapping to the list of UIs.
    - If I run `:detach`, how will the server know which UI is the "current UI"?
    - Or a more general solution is to provide a way 
3. the builtin UI should call `nvim_set_client_info` to provide all possible info about itself, including PID.
    - but that still doesn't solve (2).

**Reply by LunarLambda** [reply]
Mmhh. I see the crux of the issue now. I'm not sure I have an immediate suggestion for how to approach it, however.

If I understand correctly:

1. `nvim_get_chan_info(0)` can never return a useful result if called from Lua/Vimscript, which always executes in the "server" context, so any client that wants to expose it to the user has to have its own integration/plugin functionality.
2. The UI doesn't, for example, wait for the user to finish typing `:foo<cr>` such that the server could receive an atomic message like "UI `#3` sent command :foo", such that `:foo` could make use of UI-specific information, instead it just receives individual keypress events from every UI, so outside of RPC messages it's hard to express a concrete, UI-specific intention to the server.

It's indeed quite difficult to design a feature for as every single `nvim` TUI client is currently indistinguishable from each other in terms of its observable and interactable state.

**Reply by justinmk** [reply]
I think the ui can look for something like `[count]<c-z>` and handle that. Workin on a PR.

**Reply by LunarLambda** [reply]
Interesting, what would the count be for?
detach UI N, with 0 = "current"?

**Reply by justinmk** [reply]
to distinguish from normal ctrl-z "backgrounding" behavior (presumably there is a use-case for that)

**Reply by LunarLambda** [reply]
> For the default case where the nvim ui spawns the server as a _child_ process, I think it should work if (1) the `nvim` server is spawned with `detach=true`, (2) terminating the _parent_ (ui) process should allow the _child_ (server) to continue running.

How would re-attach work in that case? The default Neovim server uses `nvim --embed` which uses stdio. If the parent (ui) process goes away, then there is nothing that "remembers" that embedded process and could reconnect to it.

Maybe Ctrl-Z should be suspend if "local"/non-remote-ui is used (--embed server), and detach if remote-ui is used?

Or detach if at least 1 servername (--listen, v:servernames) exists?

**Comment by justinmk** [comment]
With https://github.com/neovim/neovim/pull/30319 , `ctrl-z` will detach the current UI (works for any UI/TUI/GUI). Please try it out.
